### DISCUSSION TITLE
E3100CB Test

### DISCUSSION THREAD

**Discussion by rtr1001** [MyHomeMyData/E3onCAN#5] [Show and tell]
Hi @MyHomeMyData,

was für ein gutes Timing, dass Du gestern die Implementierung für den E3100CB hochgeladen hast :-) An dieser Stelle meinen Dank für die Arbeit!

Ich hab mir dann gleich die Datenfelder angesehen und mit meinem E3100CB abgeglichen:

| Datenpunkt | Ergebnis | Einheit | Kommentar |
|---|---|---|---|
|1385.1|GridEnergy: {"ImportCumulated": 15361.8}|kWh|:white_check_mark:|
|1385.2|GridEnergy: {"ExportCumulated": 4695.54}|kWh|:white_check_mark:|
|1385.3|GridState: {"OperationState": -1}|1|:white_check_mark:|
|1385.4|GridActivePower: {"Total": -11.0}|W|:white_check_mark:|
|1385.5|GridReactivePower: {"Total": 302.0}|var|:white_check_mark:|
|1385.6|GridCurrent: {"Absolute L1": 1.0}|A|Das ist der Betrag der Stromstärke. Der Wert auf dem Bus **immer positiv**, auch bei negativen Strömen. Die Genauigkeit ist auch nur 1A. Falls man die Stromstärke für irgendwas braucht dann besser gleich Leistung/Spannung rechnen.|
|1385.7|GridVoltage: {"L1": 233.0}|V|:white_check_mark:|
|1385.8|GridActivePower: {"L1": 172.0}|W|:white_check_mark:|
|1385.9|GridReactivePower: {"L1": -311.0}|var|:white_check_mark:|
|1385.10|GridCurrent: {"Absolute L2": 16.0}|A|Betrag der Stromstärke in A (ohne Vorzeichen)|
|1385.11|GridVoltage: {"L2": 236.0}|V|:white_check_mark:|
|1385.12|GridActivePower: {"L2": -3737.0}|W|:white_check_mark:|
|1385.13|GridReactivePower: {"L2": 720.0}|var|:white_check_mark:|
|1385.14|GridCurrent: {"Absolute L3": 15.0}|A|Betrag der Stromstärke in A (ohne Vorzeichen)|
|1385.15|GridVoltage: {"L3": 232.0}|V|:white_check_mark:|
|1385.16|GridActivePower: {"L3": 3554.0}|W|:white_check_mark:|
|1385.17|GridReactivePower: {"L3": -107.0}|var|:white_check_mark:|

Noch eine Anmerkung zu den Blindleistungen (Reactive Power): Im Zähler selber werden diese ohne Vorzeichen angezeigt. Die Werte auf den CAN die hier dargestellt sind, sind vorzeichenbehaftet. Die Summe der drei Werte ergibt dann auch immer die totale Blindleistung.

Da ich hier noch neu bin und die Gepflogenheiten noch nicht kenne, hier meine Änderungen an der E3onCANdatapointsE3100CB.py:

```
import E3onCANcodecs
from E3onCANcodecs import *

dataIdentifiersE3100CB = {
    "1385.1": E3ComplexType(4, "GridEnergy", [E3Float32(4, "ImportCumulated", scale=1000.0)]),
    "1385.2": E3ComplexType(4, "GridEnergy", [E3Float32(4, "ExportCumulated", scale=1000.0)]),
    "1385.3": E3ComplexType(1, "GridState", [E3StateEM(1, "OperationState", offset=0)]),
    "1385.4": E3ComplexType(4, "GridActivePower", [E3Int16(2, "Total", scale=1.0, signed=True)]),
    "1385.5": E3ComplexType(4, "GridReactivePower", [E3Int16(2, "Total", scale=1.0, signed=True)]),
    "1385.6": E3ComplexType(2, "GridCurrent", [E3Int16(2, "Absolute L1", scale=1.0, signed=True)]),
    "1385.7": E3ComplexType(4, "GridVoltage", [E3Int32(4, "L1", scale=1.0, signed=False)]),
    "1385.8": E3ComplexType(4, "GridActivePower", [E3Int16(2, "L1", scale=1.0, signed=True)]),
    "1385.9": E3ComplexType(4, "GridReactivePower", [E3Int16(2, "L1", scale=1.0, signed=True)]),
    "1385.10": E3ComplexType(2, "GridCurrent", [E3Int16(2, "Absolute L2", scale=1.0, signed=True)]),
    "1385.11": E3ComplexType(4, "GridVoltage", [E3Int32(4, "L2", scale=1.0, signed=False)]),
    "1385.12": E3ComplexType(4, "GridActivePower", [E3Int16(2, "L2", scale=1.0, signed=True)]),
    "1385.13": E3ComplexType(4, "GridReactivePower", [E3Int16(2, "L2", scale=1.0, signed=True)]),
    "1385.14": E3ComplexType(2, "GridCurrent", [E3Int16(2, "Absolute L3", scale=1.0, signed=True)]),
    "1385.15": E3ComplexType(4, "GridVoltage", [E3Int32(4, "L3", scale=1.0, signed=False)]),
    "1385.16": E3ComplexType(4, "GridActivePower", [E3Int16(2, "L3", scale=1.0, signed=True)]),
    "1385.17": E3ComplexType(4, "GridReactivePower", [E3Int16(2, "L3", scale=1.0, signed=True)])
}
```

**Comment by MyHomeMyData** [comment]
@rtr1001 Wow, das ist super! Vielen Dank fürs Testen und die ausführliche Rückmeldung. 1385.5 als Reactive Power Total sieht auch mit den Log-Daten, die ich habe, gut und stimmig aus. Bei den Strömen ist mir auch schon aufgefallen, dass keine negativen Werte auftreten.

Habe Deine Vorschläge übernommen und eben nach develop gepusht. Bitte nochmal testen.

**Reply by rtr1001** [reply]
Habs getestet, läuft perfekt. Danke für die schnelle Umsetzung. Jetzt kann die Sonne kommen 😄

**Reply by Luk3la** [reply]
Hey,
kann es sein, dass der nachfolgende Teil auch für den E380 Zähler gilt? Bei mir ist die Stromstärke immer minimum 1.0A, nach oben hin ändert sich dann aber was. Bspw. wenn ich voll einspeise steht auf einer Phase dann 16A.

> GridCurrent: Das ist der Betrag der Stromstärke. Der Wert auf dem Bus immer positiv, auch bei negativen Strömen. Die Genauigkeit ist auch nur 1A. Falls man die Stromstärke für irgendwas braucht dann besser gleich Leistung/Spannung rechnen.

Gruß

**Reply by MyHomeMyData** [reply]
@Luk3la Ja, gilt auch für den E380 - siehe auch [Tabelle der Einheiten](https://github.com/MyHomeMyData/E3onCAN#e380-data-and-units).

**Reply by Luk3la** [reply]
Okay, dann habe ich das falsch interpretiert. Ich bin davon ausgegangen, dass dann bspw. 0.1A als Wert raus kommt.
Dann bleibt ja noch das gute alte URI um den Wert zu berechnen.

Gruß

**Comment by pellenbeck** [comment]
<img width="342" alt="Bildschirmfoto 2024-04-18 um 10 31 10" src="https://github.com/MyHomeMyData/E3onCAN/assets/86413310/840315bd-3926-49ac-b780-deb1e1c26b0d">

**Comment by pellenbeck** [comment]
@rtr1001 @MyHomeMyData 
Hier noch der bisherige Verlauf der Analyse: https://github.com/open3e/open3e/issues/87

**Reply by MyHomeMyData** [reply]
@pellenbeck Verwendest Du ioBroker? E3onCAN gibt's ja auch als Adapter für den ioBroker und ich habe da jetzt den E3100CB auch eingebaut. Leider kann ich das bei mir nicht im Realbetrieb testen. Wäre klasse, wenn Du das machen könntest.

Der Adapter ist im Beta-Repo verfügbar. Die Version mit E3100CB habe ich aber noch nicht released. Deshalb müsstest Du das von Github installieren. Am einfachsten geht das mit
```
iob url https://github.com/MyHomeMyData/ioBroker.e3oncan
```
Dann bei den Adaptern nach e3oncan suchen und eine Instanz erzeugen. Alles weitere siehe [Readme](https://github.com/MyHomeMyData/ioBroker.e3oncan). Ein Device-Scan ist notwendig, einen Datenpunkt-Scan musst Du nicht machen, wenn Du nur das Energy Meter testen möchtest.

@rtr1001 Falls Du auch ioBroker verwendest: Je mehr getestet wird, umso besser.

**Reply by pellenbeck** [reply]
@MyHomeMyData habe das gerade versucht zu installieren mit folgenden Fehler - kenne mich beim ioBroker bzw. der Installation von Beta-Projekten nicht aus:

```
philipp@pi-can:~/tmp-dir $ iob url https://github.com/MyHoneMyData/ioBroker.e3oncan
Info: Can not get current GitHub commit, only remember that we installed from GitHub: Request failed with status code 404
install MyHoneMyData/ioBroker.e3oncan
NPM version: 10.2.4
Installing MyHoneMyData/ioBroker.e3oncan... (System call)
npm ERR! code 128
npm ERR! An unknown git error occurred
npm ERR! command git --no-replace-objects ls-remote ssh://git@github.com/MyHoneMyData/ioBroker.e3oncan.git
npm ERR! Warning: Permanently added 'github.com,140.82.121.3' (ECDSA) to the list of known hosts.
npm ERR! git@github.com: Permission denied (publickey).
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.

npm ERR! A complete log of this run can be found in: /home/iobroker/.npm/_logs/2024-04-19T18_41_13_553Z-debug-0.log
host.pi-can Cannot install MyHoneMyData/ioBroker.e3oncan: 128
philipp@pi-can:~//tmp-dir $ cat  /home/iobroker/.npm/_logs/2024-04-19T18_41_13_553Z-debug-0.log
0 verbose cli /usr/bin/node /usr/bin/npm
1 info using npm@10.2.4
2 info using node@v18.19.1
3 timing npm:load:whichnode Completed in 6ms
4 timing config:load:defaults Completed in 11ms
5 timing config:load:file:/usr/lib/node_modules/npm/npmrc Completed in 1ms
6 timing config:load:builtin Completed in 3ms
7 timing config:load:cli Completed in 17ms
8 timing config:load:env Completed in 2ms
9 timing config:load:file:/opt/iobroker/.npmrc Completed in 13ms
10 timing config:load:project Completed in 16ms
11 timing config:load:file:/home/iobroker/.npmrc Completed in 0ms
12 timing config:load:user Completed in 1ms
13 timing config:load:file:/usr/etc/npmrc Completed in 1ms
14 timing config:load:global Completed in 1ms
15 timing config:load:setEnvs Completed in 6ms
16 timing config:load Completed in 59ms
17 timing npm:load:configload Completed in 60ms
18 timing config:load:flatten Completed in 17ms
19 timing npm:load:mkdirpcache Completed in 2ms
20 timing npm:load:mkdirplogs Completed in 1ms
21 verbose title npm install MyHoneMyData/ioBroker.e3oncan
22 verbose argv "install" "--loglevel" "error" "MyHoneMyData/ioBroker.e3oncan"
23 timing npm:load:setTitle Completed in 6ms
24 timing npm:load:display Completed in 5ms
25 verbose logfile logs-max:10 dir:/home/iobroker/.npm/_logs/2024-04-19T18_41_13_553Z-
26 verbose logfile /home/iobroker/.npm/_logs/2024-04-19T18_41_13_553Z-debug-0.log
27 timing npm:load:logFile Completed in 52ms
28 timing npm:load:timers Completed in 0ms
29 timing npm:load:configScope Completed in 1ms
30 timing npm:load Completed in 227ms
31 timing arborist:ctor Completed in 5ms
32 silly logfile start cleaning logs, removing 4 files
33 timing idealTree:init Completed in 3466ms
34 silly logfile done cleaning log files
35 verbose shrinkwrap failed to load node_modules/.package-lock.json out of date, updated: node_modules/iobroker.js-controller
36 timing idealTree Completed in 5996ms
37 timing command:install Completed in 6054ms
38 verbose stack Error: An unknown git error occurred
38 verbose stack     at makeError (/usr/lib/node_modules/npm/node_modules/@npmcli/git/lib/make-error.js:28:13)
38 verbose stack     at /usr/lib/node_modules/npm/node_modules/@npmcli/git/lib/spawn.js:37:26
39 verbose cwd /opt/iobroker
40 verbose Linux 6.1.21-v7+
41 verbose node v18.19.1
42 verbose npm  v10.2.4
43 error code 128
44 error An unknown git error occurred
45 error command git --no-replace-objects ls-remote ssh://git@github.com/MyHoneMyData/ioBroker.e3oncan.git
46 error Warning: Permanently added 'github.com,140.82.121.3' (ECDSA) to the list of known hosts.
46 error git@github.com: Permission denied (publickey).
46 error fatal: Could not read from remote repository.
46 error
46 error Please make sure you have the correct access rights
46 error and the repository exists.
47 verbose exit 128
48 timing npm Completed in 8231ms
49 verbose unfinished npm timer reify 1713552075721
50 verbose unfinished npm timer reify:loadTrees 1713552075755
51 verbose unfinished npm timer idealTree:userRequests 1713552079228
52 verbose code 128
53 error A complete log of this run can be found in: /home/iobroker/.npm/_logs/2024-04-19T18_41_13_553Z-debug-0.log
```

**Reply by rtr1001** [reply]
Ich nutze es nur für EVCC und habe es seit gestern produktiv am Laufen.

**Reply by pellenbeck** [reply]
@rtr1001 dafür nutze ich es auch. Welche Werte übermittelst du alle an EVCC?
Du hast nicht zufällig noch eine zweite PV/einen zweiten WR, der in den E3100CB "reinläuft"?

**Reply by rtr1001** [reply]
Ich habe nur einen WR. Ich verwende die Werte wie in den ersten beiden Beiträgen der Diskussion: https://github.com/evcc-io/evcc/discussions/10457, wobei ich für unseren Zähler den 1385.04 Wert nehme:
python3 ~/E3onCAN/E3onCANcollect.py -c can0 -dev e3100cb -r 1385.04 -m localhost:1883:open3e -retain 1385.04 &

Ich habe die Werte vom Zähler etwas beobachtet und gesehen, dass die VX3 (mit Batteriespeicher) nicht allzu schnell und auch nicht allzu gut die Leistung auf 0W regelt. Daher berechne ich aus den Zählerwerten mittlerweile einen Mittelwert über 5s und schicke den auf den MQTT Server.

**Reply by pellenbeck** [reply]
Hm, ich würde aber auch nicht direkt mit den String-Werten arbeiten. Die Leistung kommt aus dem Wechselrichter des VX3 nicht zu 100% raus; es gibt immer Leistungsverluste bzw. braucht der VX3 ja selber auch Strom etc..
Das ist aber auch quasi genau mein Problem: Es gibt keinen Wert für "nur" PV-Wechselrichter-Leistung. Nur Werte für den WR insgesamt, da sind dann aber auch schon Batteriewerte (Lade/Entladung) enthalten.

**Comment by MyHomeMyData** [comment]
Anscheinend fehlt im Link in .git. Probier's mal damit:
```
iob url https://github.com/MyHomeMyData/ioBroker.e3oncan.git
```

**Reply by pellenbeck** [reply]
@MyHomeMyData Sorry, war ein Typo bei mir "MyHoneMyData" ;-)
Ist installiert. Schaue es mir an!

**Reply by pellenbeck** [reply]
@MyHomeMyData Denke das sieht so gut aus, oder?
![Bildschirmfoto 2024-04-19 um 21 53 21](https://github.com/MyHomeMyData/E3onCAN/assets/86413310/fe3de7b1-49aa-469b-a2a2-62f5cb053d9a)

**Reply by MyHomeMyData** [reply]
@pellenbeck Ja, für mich sieht das gut aus. Die Werte sind hoffentlich auch plausibel.

Vielen Dank fürs Testen 👍

**Reply by pellenbeck** [reply]
@MyHomeMyData gerne, der Dank gilt aber dir ;-)

Auf dem Screenshot fehlen noch 2 Werte, die aber mittlerweile auch da sind bzw. geliefert werden. Und ja, plausibel sind die auch.
